Firestore/core/src/firebase/firestore/local/local_store.cc

```diff
@@ -426,53 +456,57 @@ QueryData LocalStore::AllocateQuery(Query query) {
   // Sanity check to ensure that even when resuming a query it's not currently
   // active.
   TargetId target_id = query_data.target_id();
-  HARD_ASSERT(target_ids.find(target_id) == target_ids.end(),
-              "Tried to allocate an already allocated query: %s",
-              query.ToString());
-  target_ids[target_id] = query_data;
+  HARD_ASSERT(
+      query_data_by_target_.find(target_id) == query_data_by_target_.end(),
+      "Tried to allocate an already allocated query: %s", query.ToString());
+  query_data_by_target_[target_id] = query_data;
+  target_id_by_query_[query] = target_id;
   return query_data;
 }
 
 void LocalStore::ReleaseQuery(const Query& query) {
   persistence_->Run("Release query", [&] {
-    absl::optional<QueryData> query_data = query_cache_->GetTarget(query);
+    absl::optional<QueryData> query_data = GetQueryData(query);
     HARD_ASSERT(query_data, "Tried to release nonexistent query: %s",
                 query.ToString());
 
-    TargetId target_id = query_data->target_id();
-
-    auto found = target_ids.find(target_id);
-    if (found != target_ids.end()) {
-      const QueryData& cached_query_data = found->second;
-
-      if (cached_query_data.snapshot_version() >
-          query_data->snapshot_version()) {
-        // If we've been avoiding persisting the resume_token (see
-        // ShouldPersistQueryData for conditions and rationale) we need to
-        // persist the token now because there will no longer be an in-memory
-        // version to fall back on.
-        query_data = cached_query_data;
-        query_cache_->UpdateTarget(*query_data);
-      }
-    }
-
     // References for documents sent via Watch are automatically removed when we
     // delete a query's target data from the reference delegate. Since this does
     // not remove references for locally mutated documents, we have to remove
     // the target associations for these documents manually.
-    DocumentKeySet removed = local_view_references_.RemoveReferences(target_id);
+    DocumentKeySet removed =
+        local_view_references_.RemoveReferences(query_data->target_id());
     for (const DocumentKey& key : removed) {
       persistence_->reference_delegate()->RemoveReference(key);
     }
-    target_ids.erase(target_id);
+    query_data_by_target_.erase(query_data->target_id());
+    target_id_by_query_.erase(query);
+
+    // Note: This also updates the query cache
     persistence_->reference_delegate()->RemoveTarget(*query_data);
   });
 }
 
-DocumentMap LocalStore::ExecuteQuery(const Query& query) {
+QueryResult LocalStore::ExecuteQuery(const Query& query,
+                                     bool use_previous_results) {
   return persistence_->Run("ExecuteQuery", [&] {
-    return local_documents_->GetDocumentsMatchingQuery(query,
-                                                       SnapshotVersion::None());
+    absl::optional<QueryData> query_data = GetQueryData(query);
```

Comment:
Port of https://github.com/firebase/firebase-android-sdk/blob/d8ff6a54d12fd676a31c2c331f09fe0a7d033202/firebase-firestore/src/main/java/com/google/firebase/firestore/local/LocalStore.java#L617